### PR TITLE
net/accept: alloc the accept fd after accept success

### PR DESCRIPTION
**Pull request by anchao**
## Summary

net/accept: alloc the accept fd after accept success

The new fd will be in a pending state before accept is successful, if the application creates a task when accepting, the pending fd will be dup(), but since newsock->s_conn has not been initialized, assert will be triggered when inet_addref()

`[  314.683402] [76] [ EMERG] up_assert: Assertion failed at file:inet/inet_sockif.c line: 312 task: pt-0xc19efad`

https://github.com/apache/incubator-nuttx/blob/7c20199a61f61f1bd781cebee34cdc2547841564/net/inet/inet_sockif.c#L312

## Impact

N/A

## Testing

Thread 1 : accept()
Thread 2 : posix_spawn()

### CODE REVIEW

**Comment by yamt** [comment]
> The new fd will be in a pending state before accept is successful, if the application creates a task when accepting, the pending fd will be dup(), but since newsock->s_conn has not been initialized, assert will be triggered when inet_addref()

do you mean that, while a thread is blocking in accept(), another thread in the same task group causes dup() of the fd?

**Comment by anchao** [comment]
> > The new fd will be in a pending state before accept is successful, if the application creates a task when accepting, the pending fd will be dup(), but since newsock->s_conn has not been initialized, assert will be triggered when inet_addref()
> 
> do you mean that, while a thread is blocking in accept(), another thread in the same task group causes dup() of the fd?

Yep.

**Comment by xiaoxiang781216** [comment]
It's a common case that the main thread is blocking in accept to wait the new connection, and many work threads service the client request concurrently. If the work thread decide to execute some program, the crash will happen immediately.

File: net/socket/accept.c [hunk 1]
```diff
@@ -259,29 +263,31 @@ int accept(int sockfd, FAR struct sockaddr *addr, FAR socklen_t *addrlen)
       goto errout;
     }
 
+  ret = psock_accept(psock, addr, addrlen, &newsock);
```
Comment (yamt) [inline net/socket/accept.c:266]: this doesn't look safe as accept can record the reference to newsock somewhere.
eg. tcp_start_monitor

**Review by yamt — CHANGES_REQUESTED** [review]

**Comment by yamt** [comment]
i think even socket() has the same problem potentially.

maybe it's cleaner to separate sockfd_allocate into two.
* allocate socket (just kmm_zalloc at this point)
* allocate a file and associate the given already initialized socket

**Comment by xiaoxiang781216** [comment]
> i think even socket() has the same problem potentially.
> 
> maybe it's cleaner to separate sockfd_allocate into two.
> 
> * allocate socket (just kmm_zalloc at this point)
> * allocate a file and associate the given already initialized socket

Yes, there is potential risk in other case, @Donny9 please find a general solution to fix the race condition. But this patch is valuable to avoid waste one additional file handle resource. @yamt do you think so?

File: net/socket/accept.c (hunk 1 above)
Comment (xiaoxiang781216) [reply net/socket/accept.c:266]: Good, catch.

**Comment by yamt** [comment]
> > i think even socket() has the same problem potentially.
> > maybe it's cleaner to separate sockfd_allocate into two.
> > 
> > * allocate socket (just kmm_zalloc at this point)
> > * allocate a file and associate the given already initialized socket
> 
> Yes, there is potential risk in other case, @Donny9 please find a general solution to fix the race condition. But this patch is valuable byself to avoid waste one additional file handle resource. @yamt do you think so?

i'm not against a quick fix as far as it doesn't have other issues. (like tcp_start_monitor thing in this patch)

**Comment by xiaoxiang781216** [comment]
Yes, we need resolve tcp_monitor issue first.

**Comment by anchao** [comment]
> > > i think even socket() has the same problem potentially.
> > > maybe it's cleaner to separate sockfd_allocate into two.
> > > 
> > > * allocate socket (just kmm_zalloc at this point)
> > > * allocate a file and associate the given already initialized socket
> > 
> > 
> > Yes, there is potential risk in other case, @Donny9 please find a general solution to fix the race condition. But this patch is valuable byself to avoid waste one additional file handle resource. @yamt do you think so?
> 
> i'm not against a quick fix as far as it doesn't have other issues. (like tcp_start_monitor thing in this patch)

Hi @yamt san,

In the new rebase, I moved the psock alloc outside of sockfd_allocate, please review it again

File: fs/socket/socket.c [hunk 2]
```diff
@@ -285,6 +284,9 @@ int socket(int domain, int type, int protocol)
 errout_with_sockfd:
   nx_close(sockfd);
```
Comment (yamt) [inline fs/socket/socket.c:285]: doesn't this free the psock and causes double free below?

File: net/socket/accept.c [hunk 3]
```diff
@@ -259,29 +260,36 @@ int accept(int sockfd, FAR struct sockaddr *addr, FAR socklen_t *addrlen)
...
+      goto errout_with_alloc;
     }
 
   leave_cancellation_point();
   return newfd;
 
-errout_with_socket:
-  nx_close(newfd);
+errout_with_alloc:
+  kmm_free(newsock);
```
Comment (yamt) [inline net/socket/accept.c:292]: i suspect you should use psock_close here.

**Review by yamt — CHANGES_REQUESTED** [review]

File: fs/socket/socket.c (hunk 2 above)
Comment (anchao) [reply fs/socket/socket.c]: Done

File: net/socket/accept.c (hunk 3 above)
Comment (anchao) [reply net/socket/accept.c:295]: Done

**Review by xiaoxiang781216 — APPROVED** [review]
LGTM, please take a look @yamt

**Review by yamt — APPROVED** [review]
lgtm

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**